### PR TITLE
[compiler][fixtures] test repros: codegen, alignScope, phis

### PR DESCRIPTION
**Pull request by mofeiZ**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30078
* #30077
* #30076
* #30075
* #30074
* #29891
* __->__ #29878

The AlignReactiveScope bug should be simplest to fix, but it's also caught by an invariant assertion. I think a fix could be either keeping track of "active" block-fallthrough pairs (`retainWhere(pair => pair.range.end > current.instr[0].id)`) or following the approach in `assertValidBlockNesting`.
I'm tempted to pull the value-block aligning logic out into its own pass (using the current `node` tree traversal), then align to non-value blocks with the `assertValidBlockNesting` approach. Happy to hear feedback on this though!

The other two are likely bigger issues, as they're not caught by static invariants.

Update:
- removed bug-phi-reference-effect as it's been patched by @josephsavona
- added bug-array-concat-should-capture

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-codegen-inline-iife.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ *   logs: [2,1]
+ */
+function useTest() {
+  return makeArray<number | void>(
+    print(1),
+    (function foo() {
+      print(2);
+      return 2;
+    })()
+  );
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-codegen-inline-iife.ts:30]: @josephsavona @mvitousek Curious as to your thoughts on this issue.

My intuition (could definitely be far off the mark) is that we can benefit from assertions in `codegen`. E.g. we can assert that generated IR instructions are used / inlined in some increasing order (while accounting for re-orderable instructions like property loads).

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-phi-reference-effect.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ *   (kind: ok) [2]
+ *   [2,2]
+ */
+function Foo(cond) {
+  let x = null;
+  if (cond) {
+    x = [];
+  } else {
+  }
+  // Here, x = phi(x$null, x$[]) does not receive the correct ValueKind
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-phi-reference-effect.ts:19]: fixed in our pair coding session, landed in #29879

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-codegen-inline-iife.ts (hunk 1 above)
Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-codegen-inline-iife.ts:30]: Per discussion: i think what we need to do here is add a compiler pass that forces temporaries to be used wherever necessary to retain instruction ordering. We're basically looking for any instruction whose lvalue is first referenced after some intervening lvalue-less instruction.

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-array-concat-should-capture.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { mutate } from "shared-runtime";
+
+/**
+ * Fixture showing why `concat` needs to capture both the callee and rest args.
+ * Here, observe that arr1's values are captured into arr2.
+ *  - Later mutations of arr2 may write to values within arr1.
+ *  - Observe that it's technically valid to separately memoize the array arr1
+ *    itself.
+ */
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-array-concat-should-capture.ts:9]: nice! This should be trivial to fix, our definition for concat() just needs to be more pessimistic:

https://github.com/facebook/react/blob/2ba462b665bb4691067e67db3cfbffc56959b788/compiler/packages/babel-plugin-react-compiler/src/HIR/ObjectShape.ts#L229-L238

**Review by josephsavona — APPROVED** [review]
Great finds!

**Comment by josephsavona** [comment]
> The other two are likely bigger issues, as they're not caught by static invariants.

Per comment the concat one isn't a general issue, it's just a misdefinition for concat.

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-array-concat-should-capture.ts (hunk 3 above)
Comment (mofeiZ) [reply compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/bug-array-concat-should-capture.ts:9]: Yep, already in my local bugfix stack!

**Review by josephsavona — APPROVED** [review]
awesome!

**Pull request merged**